src/Evolution/Systems/NewtonianEuler/Characteristics.cpp

```diff
@@ -41,6 +48,354 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
+#ifdef SPECTRE_DEBUG
+  ASSERT(equal_within_roundoff(get(magnitude(unit_normal)), 1.),
+         "Expected unit normal, but got normal with magnitude "
+             << get(magnitude(unit_normal)));
+#endif  // ifdef SPECTRE_DEBUG
+
+  const double u = get<0>(velocity);
+  const double v = get<1>(velocity);
+  const double n_x = get<0>(unit_normal);
+  const double n_y = get<1>(unit_normal);
+  const double c = sqrt(get(sound_speed_squared));
+  const double v_n = get(dot_product(velocity, unit_normal));
+
+  Matrix result(4, 4);
+  result(0, 0) = 1.;
+  result(0, 1) = 0.;
+  result(0, 2) = 1.;
+  result(0, 3) = 1.;
+  result(1, 0) = u;
+  result(1, 1) = -n_y;
+  result(1, 2) = u - n_x * c;
+  result(1, 3) = u + n_x * c;
+  result(2, 0) = v;
+  result(2, 1) = n_x;
+  result(2, 2) = v - n_y * c;
+  result(2, 3) = v + n_y * c;
+  result(3, 0) = get(specific_enthalpy) -
+                 get(sound_speed_squared) / get(kappa_over_density);
+  result(3, 1) = -n_y * u + n_x * v;
+  result(3, 2) = get(specific_enthalpy) - c * v_n;
+  result(3, 3) = get(specific_enthalpy) + c * v_n;
+  return result;
+}
+
+template <>
+Matrix right_eigenvectors<3>(const tnsr::I<double, 3>& velocity,
+                             const Scalar<double>& sound_speed_squared,
+                             const Scalar<double>& specific_enthalpy,
+                             const Scalar<double>& kappa_over_density,
+                             const tnsr::i<double, 3>& unit_normal) noexcept {
+#ifdef SPECTRE_DEBUG
+  ASSERT(equal_within_roundoff(get(magnitude(unit_normal)), 1.),
+         "Expected unit normal, but got normal with magnitude "
+             << get(magnitude(unit_normal)));
+#endif  // ifdef SPECTRE_DEBUG
+
+  const double u = get<0>(velocity);
+  const double v = get<1>(velocity);
+  const double w = get<2>(velocity);
+  const double n_x = get<0>(unit_normal);
+  const double n_y = get<1>(unit_normal);
+  const double n_z = get<2>(unit_normal);
+  const double c = sqrt(get(sound_speed_squared));
+  const double v_n = get(dot_product(velocity, unit_normal));
+
+  Matrix result(5, 5);
+  result(0, 0) = 1.;
+  result(1, 0) = u;
+  result(2, 0) = v;
+  result(3, 0) = w;
+  result(4, 0) = get(specific_enthalpy) -
+                 get(sound_speed_squared) / get(kappa_over_density);
+  result(0, 3) = 1.;
+  result(1, 3) = u - n_x * c;
+  result(2, 3) = v - n_y * c;
+  result(3, 3) = w - n_z * c;
+  result(4, 3) = get(specific_enthalpy) - v_n * c;
+  result(0, 4) = 1.;
+  result(1, 4) = u + n_x * c;
+  result(2, 4) = v + n_y * c;
+  result(3, 4) = w + n_z * c;
+  result(4, 4) = get(specific_enthalpy) + v_n * c;
+
+  // There is some degeneracy in the choice of the column eigenvectors. The row
+  // eigenvectors are chosen so that the largest of (n_x, n_y, n_z) appears in
+  // the denominator, as this avoids division by zero. Here we make the
+  // consistent choice of right eigenvectors.
+  const size_t index_of_largest = static_cast<size_t>(
```

Comment:
since you're never using `index_of_largest` to index into a data structure you could do `const auto ... = std::dist` to avoid the cast. Up to you :)